app/scripts/migrations/127.ts

```diff
@@ -0,0 +1,414 @@
+    // configuration for this chain id. The network controller requires
+    // configurations to have at least 1 endpoint.
+    if (rpcEndpoints.length === 0) {
+      return acc;
+    }
+
+    // Use the tie breaker network as the default rpc endpoint
+    const defaultRpcEndpointIndex = Math.max(
+      rpcEndpoints.findIndex(
+        (endpoint) => endpoint.networkClientId === tieBreaker?.id,
+      ),
+      // Or arbitrarily default to the first endpoint if we don't have a tie breaker
+      0,
+    );
+
+    // Calculate the unique array of non-empty block explorer urls
+    const blockExplorerUrls = [
+      ...networks.reduce((urls, network) => {
+        if (
+          isObject(network.rpcPrefs) &&
+          typeof network.rpcPrefs.blockExplorerUrl === 'string' &&
+          network.rpcPrefs.blockExplorerUrl
+        ) {
+          urls.add(network.rpcPrefs.blockExplorerUrl);
+        }
+        return urls;
+      }, new Set()),
+    ];
+
+    // Use the tie breaker network as the default block explorer, if it has one
+    const defaultBlockExplorerUrlIndex =
+      blockExplorerUrls.length === 0
+        ? undefined
+        : Math.max(
+            blockExplorerUrls.findIndex(
+              (url) =>
+                isObject(tieBreaker?.rpcPrefs) &&
+                url === tieBreaker.rpcPrefs.blockExplorerUrl,
+            ),
+            // Or arbitrarily default to the first url
+            0,
+          );
+
+    // Use the cononical network name and currency, if we have constants for them.
+    // Otherwise prefer the tie breaker's name + currency, if it defines them.
+    // Otherwise fall back to the name + currency from arbitrary networks that define them.
+    const name =
+      NETWORK_TO_NAME_MAP[chainId as keyof typeof NETWORK_TO_NAME_MAP] ??
+      tieBreaker?.nickname ??
+      networks.find((n) => n.nickname)?.nickname;
+
+    const nativeCurrency =
+      CHAIN_ID_TO_CURRENCY_SYMBOL_MAP[
+        chainId as keyof typeof CHAIN_ID_TO_CURRENCY_SYMBOL_MAP
+      ] ??
+      tieBreaker?.ticker ??
+      networks.find((n) => n.ticker)?.ticker;
+
+    acc[chainId] = {
+      chainId,
+      rpcEndpoints,
+      defaultRpcEndpointIndex,
+      blockExplorerUrls,
+      ...(defaultBlockExplorerUrlIndex !== undefined && {
+        defaultBlockExplorerUrlIndex,
+      }),
+      name,
+      nativeCurrency,
+    };
+    return acc;
+  }, {});
+
+  // Ensure that selectedNetworkClientId points to
+  // some endpoint of some network configuration.
+  let selectedNetworkClientId = Object.values(networkConfigurationsByChainId)
+    .flatMap((n) =>
+      isObject(n) && Array.isArray(n.rpcEndpoints) ? n.rpcEndpoints : [],
+    )
+    .find(
+      (e) => e.networkClientId === networkState.selectedNetworkClientId,
+    )?.networkClientId;
+
+  // It may not, if its endpoint was not well formed
+  // or a duplicate. In that case, fallback to mainnet
+  if (!selectedNetworkClientId) {
```

Comment:
Yes, this can attempt to leave them on the new default endpoint for the same chain as before, as long as we didn't have to remove the entire chain due to invalidity, in which case you may still have to default to somewhere.

Similar feedback from Jiexi: https://github.com/MetaMask/metamask-extension/pull/26433#discussion_r1755139780

Reply:
I improved this in https://github.com/MetaMask/metamask-extension/pull/26433/commits/396c47c8579f61488c26e6aba13a238ebff4c9c8

Now if `selectedNetworkClientId` points to some invalid URL, it will fall back to the new default on the same chain if possible, rather than always falling back to mainnet